File: game.js

var game = (function () {
    "use strict";

    function getInitialState () {
        return {
            paddles: [0, 0],
            ball: {x:0, y:0, vx:0.01, vy:0.005}
        };
    }

    function stepper (inputs, state)
    {
        var ns = {
            paddles: [
                state.paddles[0] + 0.05*inputs[0].paddle,
                state.paddles[1] + 0.05*inputs[1].paddle
            ],
            ball: {
                 x: state.ball.x + state.ball.vx,
                 y: state.ball.y + state.ball.vy,
                vx: state.ball.vx,
                vy: state.ball.vy
            }
        };

        if (ns.ball.x < -1) ns.ball.vx =  Math.abs (ns.ball.vx);
        if (ns.ball.x >  1) ns.ball.vx = -Math.abs (ns.ball.vx);
        if (ns.ball.y < -1) ns.ball.vy =  Math.abs (ns.ball.vy);
        if (ns.ball.y >  1) ns.ball.vy = -Math.abs (ns.ball.vy);

        return ns;
    }

    return {
        init: getInitialState,
        step: stepper,
        dt: 33
    };

})();

File: example/fay/Game.hs
module Game (
    dt,
    players,
    initialState,
    defaultInput,
    step
) where


data Ball = Ball {
    x :: Double
  , y :: Double
  , vx :: Double
  , vy :: Double
} deriving (Show)

data State = State {
    paddles :: (Double,Double)
  , score :: (Int,Int)
  , ball :: Ball
} deriving (Show)

data Input = Input {
    paddle :: Double
} deriving (Show)


initialState :: State
initialState = State (0.1,0) (0,0) (Ball 0.1 0 0.03 0.025)

defaultInput :: Input
defaultInput = Input 0

dt :: Int
dt = 50

players :: Int
players = 2

step :: Input -> State -> State
step i s = s { ball = stepBall (ball s) }
  where
    stepBall (Ball x y vx vy) = Ball (x+vx) (y+vy) vx vy

